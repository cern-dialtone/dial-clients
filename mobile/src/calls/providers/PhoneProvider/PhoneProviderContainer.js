import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import {
  connectionActions,
  callActions,
  recentCallsActions,
  authActions,
  authActionFactory
} from 'dial-core';
import { PhoneProvider } from './PhoneProvider';
import { phoneService } from './PhoneService';

export function mapStateToProps({ calls, auth }) {
  return {
    authToken: auth.authToken,
    doNotDisturb: calls.status.doNotDisturb,
    call: calls.call,
    toneToken: auth.toneToken
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      ...authActions,
      logout: authActionFactory(process.env.REACT_APP_API_ENDPOINT).logout,
      ...connectionActions,
      ...callActions,
      ...recentCallsActions
    },
    dispatch
  );
}

export default phoneService(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(PhoneProvider)
);
