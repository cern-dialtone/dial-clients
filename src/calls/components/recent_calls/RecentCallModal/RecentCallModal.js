import React from 'react';
import { Modal, Icon } from 'semantic-ui-react';
import moment from 'moment';
import PropTypes from 'prop-types';
import UserPhoneNumberButtonContainer from 'calls/components/UserPhoneNumberButton/UserPhoneNumberButtonContainer';

function RecentCallDuration({ endTime, startTime }) {
  if (endTime && startTime) {
    return (
      <li style={{ margin: 10 }}>
        <Icon name="stopwatch" aria-label="Call duration icon" />
        {moment.duration(moment(endTime).diff(moment(startTime))).humanize()}
      </li>
    );
  }
  return null;
}

RecentCallDuration.propTypes = {
  endTime: PropTypes.number,
  startTime: PropTypes.number
};

RecentCallDuration.defaultProps = {
  endTime: null,
  startTime: null
};

function RecentCallTime({ endTime, incoming, missed }) {
  const message = `${incoming ? `Incoming` : `Outgoing`} and ${
    missed ? `missed` : `picked up`
  } call`;

  return (
    <li style={{ margin: 10 }}>
      <Icon
        aria-label={message}
        name={incoming ? 'arrow down' : 'arrow up'}
        color={missed ? 'red' : 'green'}
      />
      {moment(endTime).calendar()}
    </li>
  );
}

RecentCallTime.propTypes = {
  endTime: PropTypes.number.isRequired,
  incoming: PropTypes.bool.isRequired,
  missed: PropTypes.bool.isRequired
};

function RecentCallModal({ selectedCall, open, handleClose }) {
  if (!selectedCall) {
    return null;
  }

  const {
    name,
    phoneNumber,
    startTime,
    endTime,
    missed,
    incoming
  } = selectedCall;

  return (
    <Modal
      // closeIcon
      size="tiny"
      open={open}
      key={`${name}-${endTime}-modal`}
      onClose={handleClose}
      data-testid="RecentCallModal"
      closeIcon={<Icon aria-label="Close call details modal" name="close" />}
    >
      <Modal.Header>
        <Icon name="phone" color="blue" /> Call details
      </Modal.Header>
      <Modal.Content>
        <Modal.Description>
          <Modal.Header></Modal.Header>
          <li style={{ margin: 10 }}>
            <Icon name="phone" />
            {phoneNumber}
          </li>
          <RecentCallDuration endTime={endTime} startTime={startTime} />
          <RecentCallTime
            endTime={endTime}
            incoming={incoming}
            missed={missed}
          />
        </Modal.Description>
        <UserPhoneNumberButtonContainer
          phoneNumber={phoneNumber}
          icon="phone"
          callBack={handleClose}
        />
      </Modal.Content>
    </Modal>
  );
}

RecentCallModal.propTypes = {
  selectedCall: PropTypes.shape({
    name: PropTypes.string.isRequired,
    phoneNumber: PropTypes.string.isRequired,
    startTime: PropTypes.number,
    endTime: PropTypes.number,
    missed: PropTypes.bool,
    incoming: PropTypes.bool
  }),
  open: PropTypes.bool,
  handleClose: PropTypes.func
};

RecentCallModal.defaultProps = {
  selectedCall: null,
  open: false,
  handleClose: () => {}
};

export default RecentCallModal;
