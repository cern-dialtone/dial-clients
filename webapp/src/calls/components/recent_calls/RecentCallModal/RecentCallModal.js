import React, { Component } from 'react';
import { Modal, Header, Icon } from 'semantic-ui-react';
import moment from "moment";
import UserPhoneNumberButtonContainer from 'calls/components/UserPhoneNumberButton/UserPhoneNumberButtonContainer';

export class RecentCallModal extends Component {
  inlineStyle = {
    modal: {
      marginTop: '0px !important',
      marginLeft: 'auto',
      marginRight: 'auto'
    }
  };
  
  getDuration(endTime, startTime) {
    if (endTime && startTime)
      return <li style={{ margin: 10 }}><Icon name="stopwatch"/>{moment.duration(moment(endTime).diff(moment(startTime))).humanize()}</li>;          
  }

  getTime(endTime, incoming, missed) {
    let Way;

    Way = <Icon name={(incoming) ? "arrow down" : "arrow up"} color={(missed) ? "red" : "green"}/>;
    return (<li style={{ margin: 10 }}>{Way}{moment(endTime).calendar()}</li>);
  }

  render() {
    const {
      name,
      phoneNumber,
      startTime,
      endTime,
      missed,
      incoming
    } = this.props.recentCall;
    const { trigger } = this.props;
    
    return (
      <Modal size="tiny" trigger={trigger} key={name+"-"+endTime+"-modal"}>
        <Modal.Header>
      <Icon name="phone" color="blue" /> Call details
        </Modal.Header>
        <Modal.Content>
          <Modal.Description>
            <Modal.Header></Modal.Header>
            <li style={{ margin: 10 }}><Icon name="phone"/>{phoneNumber}</li>
            {this.getDuration(endTime, startTime)}
            {this.getTime(endTime, incoming, missed)}
          </Modal.Description>
          <UserPhoneNumberButtonContainer
                phoneNumber={phoneNumber}
                icon="phone"
              />
        </Modal.Content>
      </Modal>
    );
  }
}

export default RecentCallModal;
