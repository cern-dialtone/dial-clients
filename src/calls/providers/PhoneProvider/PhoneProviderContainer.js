import { connect } from 'react-redux';

import { bindActionCreators } from 'redux';

import {
  callActions,
  connectionActions,
  recentCallsActions,
  authActions
} from 'dial-core';

import { info, success, warning } from 'common/actions/notifications';
import PhoneProvider from 'calls/providers/PhoneProvider/PhoneProvider';
import { withPhoneService } from 'calls/providers/PhoneProvider/PhoneService';

import dialBackendApi from 'services/api';

export function mapStateToProps({ auth, call, status, connection }) {
  return {
    doNotDisturb: status.doNotDisturb,
    call,
    tempRemote: call.tempRemote,
    authToken: auth.authToken,
    connected: connection.connected
  };
}

export function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      ...authActions,
      logout: dialBackendApi().logout,
      ...connectionActions,
      ...callActions,
      ...recentCallsActions,
      success,
      info,
      warning
    },
    dispatch
  );
}

export default withPhoneService(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(PhoneProvider)
);
