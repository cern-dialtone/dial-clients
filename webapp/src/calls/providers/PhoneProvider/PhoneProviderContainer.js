import { connect } from "react-redux";

import { bindActionCreators } from "redux";
import * as connectionActionCreators from "calls/actions/connection";
import * as callActionCreators from "calls/actions/call";
import * as recentActionCreators from "calls/actions/recent";
import * as authActionCreators from "auth/actions/auth";
import { info, success, warning } from "common/actions/notifications";
import { phoneService } from "calls/providers/PhoneProvider/PhoneProvider";
import PhoneProvider from "calls/providers/PhoneProvider/PhoneProvider";

export function mapStateToProps({ calls, auth }) {
  return {
    recipient: calls.call ? calls.call.recipient : undefined,
    onCall: calls.call ? calls.call.onCall : false,
    token: auth.token,
    doNotDisturb: calls.status.doNotDisturb
  };
}

export function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      ...connectionActionCreators,
      ...callActionCreators,
      ...recentActionCreators,
      ...authActionCreators,
      success,
      info,
      warning
    },
    dispatch
  );
}

export default phoneService(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(PhoneProvider)
);
