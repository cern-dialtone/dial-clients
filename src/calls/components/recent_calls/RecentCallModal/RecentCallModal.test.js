import React from 'react';
import PhoneProviderContainer from 'calls/providers/PhoneProvider/PhoneProviderContainer';
import { Provider } from 'react-redux';

import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

import { createHashHistory } from 'history';
import { routerMiddleware } from 'connected-react-router';
import { render, waitForElement, fireEvent } from '@testing-library/react';
import RecentCallModal from './RecentCallModal';

global.remote = {
  require: jest.fn()
};

const history = createHashHistory();
const middlewares = [thunk, routerMiddleware(history)];
const mockStore = configureMockStore(middlewares);

describe('RecentCallModal tests', () => {
  let store;
  let initialState;

  beforeEach(() => {
    initialState = {
      status: {
        doNotDisturb: false
      },
      call: {
        tempRemote: null
      },
      auth: {
        authToken: null
      },
      settings: {
        devices: []
      }
    };
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  it('renders the modal closed', async () => {
    const recentCalls = [
      {
        name: 'John Doe',
        endTime: 12345,
        phoneNumber: '222333',
        missed: true,
        incoming: true,
        startTime: 45678
      }
    ];
    const { asFragment, queryByTestId } = render(
      <RecentCallModal selectedCall={recentCalls[0]} open={false} />
    );
    const modal = queryByTestId('RecentCallModal');
    expect(modal).toBeNull();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the modal open', async () => {
    const recentCalls = [
      {
        name: 'John Doe',
        endTime: 12345,
        phoneNumber: '222333',
        missed: true,
        incoming: true,
        startTime: 45678
      }
    ];
    const { getByTestId } = render(
      <Provider store={store}>
        <PhoneProviderContainer>
          <RecentCallModal selectedCall={recentCalls[0]} open />
        </PhoneProviderContainer>
      </Provider>
    );
    await waitForElement(() => getByTestId('RecentCallModal'));
  });

  it('renders the modal open without startTime', async () => {
    const recentCalls = [
      {
        name: 'John Doe',
        endTime: 12345,
        phoneNumber: '222333',
        missed: true,
        incoming: true
      }
    ];
    const { getByTestId, getByLabelText } = render(
      <Provider store={store}>
        <PhoneProviderContainer>
          <RecentCallModal selectedCall={recentCalls[0]} open />
        </PhoneProviderContainer>
      </Provider>
    );
    await waitForElement(() => getByTestId('RecentCallModal'));
    await waitForElement(() => getByLabelText('Incoming and missed call'));
  });

  it('renders the modal open not missed not incoming', async () => {
    const recentCalls = [
      {
        name: 'John Doe',
        endTime: 12345,
        phoneNumber: '222333',
        missed: false,
        incoming: false,
        startTime: 45678
      }
    ];
    const {
      getByTestId,
      asFragment,
      getByText,
      getAllByText,
      getByLabelText
    } = render(
      <Provider store={store}>
        <PhoneProviderContainer>
          <RecentCallModal selectedCall={recentCalls[0]} open />
        </PhoneProviderContainer>
      </Provider>
    );
    await waitForElement(() => getByTestId('RecentCallModal'));
    await waitForElement(() => getAllByText('222333'));
    await waitForElement(() => getByText('a few seconds'));
    await waitForElement(() => getByText('Call details'));
    await waitForElement(() => getByLabelText('Call duration icon'));
    await waitForElement(() => getByLabelText('Outgoing and picked up call'));
    expect(asFragment()).toMatchSnapshot();
  });
});
