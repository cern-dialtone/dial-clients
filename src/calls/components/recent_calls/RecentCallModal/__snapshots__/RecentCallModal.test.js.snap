// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentCallModal tests renders the modal closed 1`] = `<DocumentFragment />`;

exports[`RecentCallModal tests renders the modal open not missed not incoming 1`] = `<DocumentFragment />`;
